pkg/pillar/cmd/domainmgr/containerd.go

```diff
@@ -0,0 +1,199 @@
+package domainmgr
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/utils"
+	uuid "github.com/satori/go.uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// containerd socket
+	ctrdSocket = "/run/containerd/containerd.sock"
+	// CtrdServicesNamespace containerd namespace for running containers
+	CtrdServicesNamespace = "eve-user-apps"
+
+	// root path to all containers
+	containersRoot = "/persist/rkt/pods/prepared"
+	// relative path to rootfs for an individual container
+	containerRootfsPath = "stage1/rootfs/opt/stage2/runx/"
+)
+
+// KeyValue a key-value pair
+type KeyValue struct {
+	Name  string
+	Value string
+}
+
+// getContainerPath return the path to the root of the container. This is *not*
+// necessarily the rootfs, which may be a layer below
+func getContainerPath(id string) string {
+	return path.Join(containersRoot, id)
+}
+
+// getContainerRootfs return the path to the root of the container filesystem
+func getContainerRootfs(id string) string {
+	return path.Join(getContainerPath(id), containerRootfsPath)
+}
+
+// containerdLoadImageTar load an image tar into the containerd content store
+func containerdLoadImageTar(filename string) (map[string]string, error) {
+	// load the content into the containerd content store
+	ctr, err := containerd.New(ctrdSocket)
```

Comment:
How does this work in terms of re-opening connections to containerd? Does the containerd.New take care of caching the open socket so that when this function gets called again it won't re-open the socket (thus leaking the previous one)? Or do we have to do this once (during initialization of domainmanager and keep ctr globally available?

Reply:
We won't leak, because we have the next few lines:

```go
	ctr, err := containerd.New(ctrdSocket)
	if err != nil {
		log.Errorf("could not connect to containerd socket at %s: %+s", ctrdSocket, err.Error())
		return nil, err
	}
	defer ctr.Close()
```

We close the file handle with a `defer`, so that shouldn't be an issue.

That having been said, I don't see any reason why we cannot create a `containerd.Client` during domainmgr initialization - or at least, check for one in a var and if it is nil, create and cache it. 

It shouldn't have any memory management impact (positive or negative), since we close it each time, and the overhead of opening/closing a grpc connection is tiny compared to actually loading the data, but it still is a good idea. Save it for a future improvement, methinks.